packages/libraries/router/src/persisted_documents.rs

```diff
@@ -0,0 +1,790 @@
+            .build()];
+
+        router::Response::error_builder()
+            .errors(errors)
+            .status_code(StatusCode::OK)
+            .context(ctx)
+            .build()
+            .unwrap()
+    }
+}
+
+impl PersistedDocumentsManager {
+    fn new(config: &Config) -> Self {
+        let agent = reqwest::Client::builder()
+            .danger_accept_invalid_certs(config.accept_invalid_certs.unwrap_or(false))
+            .connect_timeout(Duration::from_secs(config.connect_timeout.unwrap_or(5)))
+            .timeout(Duration::from_secs(config.request_timeout.unwrap_or(15)))
+            .build()
+            .expect("Failed to create reqwest client");
+
+        let cache_size = config.cache_size.unwrap_or(1000);
+        let cache = Arc::new(Mutex::new(LruCache::<String, String>::new(
+            NonZeroUsize::new(cache_size).unwrap(),
+        )));
+
+        Self {
+            agent,
+            cache,
+            config: config.clone(),
+        }
+    }
+
+    /// Extracts the document id from the request body.
+    /// In case of a parsing error, it returns the error.
+    /// This will also try to parse other GraphQL-related (see `original_req`) fields in order to 
+    /// pass it to the next layer.
+    fn extract_document_id(
+        body: &hyper::body::Bytes,
+    ) -> Result<ExpectedBodyStructure, PersistedDocumentsError> {
+        serde_json::from_slice::<ExpectedBodyStructure>(&body)
+            .map_err(PersistedDocumentsError::FailedToParseBody)
+    }
+
+    /// Resolves the document from the cache, or from the CDN
+    async fn resolve_document(&self, document_id: &str) -> Result<String, PersistedDocumentsError> {
+        let cached_record = self.cache.lock().await.get(document_id).cloned();
+
+        match cached_record {
+            Some(document) => {
+                debug!("Document {} found in cache: {}", document_id, document);
+
+                return Ok(document);
+            }
+            None => {
+                debug!(
+                    "Document {} not found in cache. Fetching from CDN",
+                    document_id
+                );
+                let cdn_document_id = str::replace(document_id, "~", "/");
+                let cdn_artifact_url = format!("{}/apps/{}", self.config.endpoint.as_ref().unwrap(), cdn_document_id);
+                info!(
+                    "Fetching document {} from CDN: {}",
+                    document_id, cdn_artifact_url
+                );
+                let cdn_response = self
+                    .agent
+                    .get(cdn_artifact_url)
+                    .header("X-Hive-CDN-Key", self.config.key.as_ref().unwrap())
+                    .send()
+                    .await;
+
+                match cdn_response {
+                    Ok(response) => {
+                        if response.status().is_success() {
+                            let document = response.text().await.unwrap();
```

Comment:
graceful error handling:

```suggestion
                            let document = response.text().await.map_err(PersistedDocumentsError::FailedToReadBody)?;
```

Reply:
good point, fixed